rust/src/gildedrose.rs

```diff
@@ -88,12 +109,208 @@ impl GildedRose {
+        pub name: String,
+        pub rounds: i32,
+        pub input: Vec<Item>,
+        pub expected: Vec<Item>,
+    }
+
+    impl TestData {
+        pub fn new(name: impl Into<String>, rounds: i32, input: Vec<Item>, expected: Vec<Item>) -> TestData {
+            TestData {
+                name: name.into(),
+                rounds,
+                input,
+                expected,
+            }
+        }
+    }
+
+    fn run_rounds(input: &mut GildedRose, rounds: i32) {
+        for _ in 0..rounds {
+            input.update_quality();
+        }
+    }
+
+    fn run_tests(tests: Vec<TestData>) {
+        for test in tests {
+            println!("{:#?}", test.name);
+            let mut rose = GildedRose::new(test.input);
+            run_rounds(&mut rose, test.rounds);
+            assert_eq!(GildedRose::new(test.expected), rose);
+        }
+    }
+
     #[test]
-    pub fn foo() {
-        let items = vec![Item::new("foo", 0, 0)];
-        let mut rose = GildedRose::new(items);
-        rose.update_quality();
+    pub fn main() {
+        let test_cases = vec![
+            TestData::new(
+                "default data set, 1 round",
+                1,
+                vec![
+                    Item::new("+5 Dexterity Vest", 10, 20),
+                    Item::new("Aged Brie", 2, 0),
+                    Item::new("Elixir of the Mongoose", 5, 7),
+                    Item::new("Sulfuras, Hand of Ragnaros", 0, 80),
+                    Item::new("Backstage passes to a TAFKAL80ETC concert", 15, 20),
+                ],
+                vec![
+                    Item::new("+5 Dexterity Vest", 9, 19),
+                    Item::new("Aged Brie", 1, 1),
+                    Item::new("Elixir of the Mongoose", 4, 6),
+                    Item::new("Sulfuras, Hand of Ragnaros", 0, 80),
+                    Item::new("Backstage passes to a TAFKAL80ETC concert", 14, 21),
+                ],
+            ),
+            TestData::new(
+                "default data set, 10 rounds",
+                10,
+                vec![
+                    Item::new("+5 Dexterity Vest", 10, 20),
+                    Item::new("Aged Brie", 2, 0),
+                    Item::new("Elixir of the Mongoose", 5, 7),
+                    Item::new("Sulfuras, Hand of Ragnaros", 0, 80),
+                    Item::new("Backstage passes to a TAFKAL80ETC concert", 15, 20),
+                ],
+                vec![
+                    Item::new("+5 Dexterity Vest", 0, 10),
+                    Item::new("Aged Brie", -8, 18),
+                    Item::new("Elixir of the Mongoose", -5, 0),
+                    Item::new("Sulfuras, Hand of Ragnaros", 0, 80),
+                    Item::new("Backstage passes to a TAFKAL80ETC concert", 5, 35),
+                ],
+            ),
+            TestData::new(
+                "default data set, 20 rounds",
+                20,
+                vec![
+                    Item::new("+5 Dexterity Vest", 10, 20),
```

Comment:
These tests are pretty easy to read and describe what they are for nicely. The only thing I would say is that I would prefer a structure where the inputs and expected outputs can be defined side by side.